command.go

```diff
@@ -756,6 +759,25 @@ func (c *Command) ValidateArgs(args []string) error {
 	return c.Args(c, args)
 }
 
+func (c *Command) ValidateRequiredFlags() error {
```

Comment:
Does this need to be a public function? Do we need to extend the API of Cobra for this?

I believe there are times when 1 of 2 flags might be required. But not both. I'd have to look at the bash that is generated. This would force both of them to be present to even run. Maybe we should we not re-use BashCompOneRequiredFlag. Should it be a first class property of some object, or should we just have another name to make 'no no, it really does need to be set' (which could also be used, in turn, by the bash completion code that only requires 1 of the set)?

Reply:
> Does this need to be a public function?

Not necessary. It can be private.

> Do we need to extend the API of Cobra for this?

@eparis  No. Since we have marked a param as required, we should enforce this validation as default, right? For the API, nothing need to be changed or updated.

> I believe there are times when 1 of 2 flags might be required. But not both. I'd have to look at the bash that is generated.

Well, but this will not help enforcing the required flag. The "required" only exists in the help message. If so, why not just adding this to the description message. They did the same thing.

When marking a flag as required, users/developers will assume that `cobra` will help validate such required flags. If unset, errors should be raised. And this is exactly what #206 wants. WDYT?

> Maybe we should we not re-use BashCompOneRequiredFlag. Should it be a first class property of some object, or should we just have another name to make

Keep open on this. I don't think we need to introduce a new const value. Maybe we can rename it. But this is not safe, since other project may reuse current `BashCompOneRequiredFlag`. Just keep it as it is. WDYT?

Reply:
Let's just make this function private.

The open question is about the changing behavior. Consider a program that had:
```
cmd.Flags().Int("need-flag-1", ...)
cmd.Flags().MarkFlagRequired("need-flag-1")
cmd.Flags().Int("or-flag-2", ...)
cmd.Flags().MarkFlagRequired("or-flag-2")
```
Where the expectation is that you must set either `--need-flag-1` or you must set `--or-flag-2`. I believed kube had such a place, but I cannot find it. There appears to be one place in kube where it has `need-flag-1` AND `need-flag-2`.  `kubectl create secret docker-registry` requires both `--username` and `--password`. AND is the behavior this proposed code is correct for. Which means at least kube will be fine with this change.

However this is NOT the behavior that the bash completions implement today. They implement an OR relationship, not an AND relationship. See: https://github.com/spf13/cobra/blob/master/bash_completions.go#L169 This worries me. I don't really mind the bash being wrong, it won't break the application, but I'm concern anyone who expected the annotation to work like the bash completions do could be broken with your change. Maybe we could/should try to change the bash completions behavior to match this behavior. But that's a different issue.


Anyone who called a function called `Mark*FlagRequired()` and didn't actually want it to be required is a fool and I'm ok if we break them. Instead of worrying about those people I decided to search for anyone who used `BashCompOneRequiredFlag` directly and could have a problem with this change. That resulted in [8 hits](https://github.com/search?utf8=%E2%9C%93&q=BashCompOneRequiredFlag+-filename%3Abash_completions.go+in%3Acode+language%3Agolang&type=Code).

6 of those 8 implemented ~this function. 1 of those used the annotation for something which I don't quite understand, but this change will not break them.

[1 place](https://github.com/organicelement/ipvanish/blob/e736120447cf1013b0293d71b1fe43f13916860c/commands/ipvanish.go#L65) set the annotation on a flag directly and I'm not convinced this change would be safe/compatible for them.

@ccustine I believe you wrote/own ipvanish and since you are the only user of `BashCompOneRequiredFlag` I'm worried this change might break would you be willing to sign off here? Would this break you? Are you willing to update to accommodate this change?

Reply:
> Where the expectation is that you must set either --need-flag-1 or you must set --or-flag-2.

> AND is the behavior this proposed code is correct for.

Yes, all required flags that marked are needed. An `AND` operator.

> 6 of those 8 implemented ~this function. 1 of those used the annotation for something which I don't quite understand, but this change will not break them.

Thanks @eparis for the result. This helps us know this change is useful. It is convincible.